server/repository/src/tests.rs

```diff
@@ -1181,4 +1181,149 @@ mod repository_test {
+                A: written
+        */
+
+        let manager_a = connection_manager.clone();
+        let process_a = tokio::spawn(async move {
+            let connection = manager_a.connection().unwrap();
+            let result: Result<(), TransactionError<RepositoryError>> = connection
+                .transaction_sync(|con| {
+                    println!("A: transaction started");
+                    let repo = ItemRowRepository::new(con);
+                    let _ = repo.find_one_by_id("tx_deadlock_id")?;
+                    println!("A: read");
+                    println!("A: Sleeping for 1000ms");
+                    let start_dt = SystemTime::now();
+                    for _ in 0..2 {
+                        //2*500ms = 1000ms
+                        std::thread::sleep(core::time::Duration::from_millis(500));
+                    }
+                    //Recording sleep duration here, as if the thread is blocked by something other than sleep you should see the duration significantly greater than 1000ms
+                    let sleep_duration = SystemTime::now()
+                        .duration_since(start_dt)
+                        .expect("Time went backwards");
+                    println!("A: Slept for {:?}", sleep_duration);
+                    println!("A: writing");
+                    let _ = repo.upsert_one(&inline_init(|i: &mut ItemRow| {
+                        i.id = "tx_deadlock_id2".to_string();
+                        i.name = "name_a".to_string();
+                    }))?;
+                    println!("A: written");
+                    Ok(())
+                });
+            result
+        });
+        let manager_b = connection_manager.clone();
+        let process_b = tokio::spawn(async move {
+            //Wait for process a to get a transaction started
+            let connection = manager_b.connection().unwrap();
+            // let _ = a_finished_reading_receiver.await;
+            println!("B: Ready to start transaction");
+            // println!("Starting transaction in blocking thread...");
+            let result: Result<(), TransactionError<RepositoryError>> = connection
+                .transaction_sync(|con| {
+                    println!("B: transaction started");
+                    let repo = ItemRowRepository::new(&con);
+                    let _ = repo.find_one_by_id("tx_deadlock_id")?;
+                    println!("B: read");
+
+                    let _ = repo.upsert_one(&inline_init(|i: &mut ItemRow| {
+                        i.id = "tx_deadlock_id".to_string();
+                        i.name = "name_b".to_string();
+                    }))?;
+                    println!("B: write 1");
+
+                    let _ = repo.upsert_one(&inline_init(|i: &mut ItemRow| {
+                        i.id = "tx_deadlock_id".to_string();
+                        i.name = "name_b_2".to_string();
+                    }))?;
+                    println!("B: write 2");
```

Comment:
Why do we need the second write?

nit: println!() -> log!() to make the test less noisy (think you have to run `cargo test -- --nocapture` to see the logs; untested)

Can the sleep time be reduced to shorten the test? The channel approach should work as now, shoudln't it?

Reply:
> Why do we need the second write?
We probably don't but doesn't hurt?

>  println!() -> log!() to make the test less noisy
println!() doesn't show in a normal test run, you need to use --nocapture to use. Couldn't figure how to get log::warn! to output in a test case, and I think it's important to easily see the output if someone is debugging this test in future.

> Can the sleep time be reduced to shorten the test? The channel approach should work as well now, shouldn't it?
Sleep time reduced to 100ms, channels aren't easy to get working correctly with 'transaction_sync'
Reducing sleep time did uncover an intermittent bug though. The busy_time out pragma needs to be set before WAL pragma to avoid a busy_timeout during the on_acquire function.

Reply:
info! would only work if you add implementation of `info!`

https://github.com/openmsupply/open-msupply/blob/fcf8d1fd1c429b5303cc96df0c769428f90b866d/server/server/src/main.rs#L11-L12

> println!() doesn't show in a normal test run, you need to use --nocapture to use

If you run the test via rust analyser (with the play button just before test method or mod, it's by default --nocapture btw.

Reply:
Doesn't  env::set_var("RUST_LOG", "info");  hard code the server to info level? Isn't the point of env_logger is to control the log level via an environment variable at runtime?

It looks like I'd need to do something like this to enable it for tests
```
    fn init() {
        let _ = env_logger::builder().is_test(true).try_init();
    }
```
SRC: https://crates.io/crates/env_logger

Personally I'm ok sticking with println!

Reply:
Yes it's fine with println! the reason for my comment, you've mentioned `Couldn't figure how to get log::warn! to output in a test case` so i tried to explain why